lib/server_connection.ml

```diff
@@ -197,45 +198,44 @@ let set_error_and_handle ?request t error =
 let report_exn t exn =
   set_error_and_handle t (`Exn exn)
 
-let advance_request_queue_if_necessary t =
-  if is_active t then begin
-    let reqd = current_reqd_exn t in
-    if Reqd.persistent_connection reqd then begin
-      if Reqd.is_complete reqd then begin
-        ignore (Queue.take t.request_queue);
-        if not (Queue.is_empty t.request_queue)
-        then t.request_handler (current_reqd_exn t);
-        wakeup_reader t;
-      end
-    end else begin
-      ignore (Queue.take t.request_queue);
-      Queue.iter Reqd.close_request_body t.request_queue;
-      Queue.clear t.request_queue;
-      Queue.push reqd t.request_queue;
-      wakeup_writer t;
-      if Reqd.is_complete reqd
-      then shutdown t
-      else if not (Reqd.requires_input reqd)
-      then shutdown_reader t
-    end
-  end else if Reader.is_closed t.reader
-  then shutdown t
-
-let _next_read_operation t =
-  advance_request_queue_if_necessary t;
-  if is_active t then begin
+let advance_request_queue t =
+  ignore (Queue.take t.request_queue);
+  if not (Queue.is_empty t.request_queue)
+  then t.request_handler (Queue.peek_exn t.request_queue);
+;;
+
+let rec _next_read_operation t =
+  if not (is_active t) then (
+    if Reader.is_closed t.reader
+    then shutdown t;
+    Reader.next t.reader
+  ) else (
     let reqd = current_reqd_exn t in
-    if      Reqd.requires_input        reqd then Reader.next t.reader
-    else if Reqd.persistent_connection reqd then `Yield
-    else begin
+    match Reqd.input_state reqd with
+    | Provide  -> Reader.next t.reader
+    | Complete -> _final_read_operation_for t reqd
+  )
+
+and _final_read_operation_for t reqd =
+  let next =
+    if not (Reqd.persistent_connection reqd) then (
       shutdown_reader t;
-      Reader.next t.reader
-    end
-  end else
-    Reader.next t.reader
+      Reader.next t.reader;
+    ) else (
+      match Reqd.output_state reqd with
+      | Wait | Consume -> `Yield
+      | Complete       ->
+        advance_request_queue t;
+        _next_read_operation t;
+    )
+  in
+  wakeup_writer t;
+  next
+;;
 
 let next_read_operation t =
   match _next_read_operation t with
+  (* XXX(dpatti): These two [`Error _] constructors are never returned *)
```

Comment:
Apologies for the noise. This looks sort of related to #164. I need to get a little more familiar with the httpaf internals to provide a patch, but i hit that issue when i accidentally sent a malformed request to an httpaf server. During the initial investigation I did notice that the `Parse error was at-least returned in an older version of the library.

Reply:
No worries! I think it is directly related to that, and I opened #184 which should address it.